### PR TITLE
Upgrade go-cmp and add new diff options to kmp

### PR DESCRIPTION
**Pull request by dgerd**
* Upgrades go-cmp to master
* Adds new kmp diff options (FieldListReporter and ShortDiffReporter)
* Update comments

FieldListReporter can be used to programatically find what fields differ
between two objects. This is useful in webhook object validation.

ShortDiffReporter was added as the default behavior for cmp.Diff is to
provide more context. The ShortDiff reporter is useful for use on API
fields where context is not necessarily as valuable. This keeps the
output short for use in log and error messages.

### CODE REVIEW

**Comment by mattmoor** [comment]
/lgtm
/approve
/hold

Am I correct that landing this will result in breaking `cmp.Diffs` to eventing / serving?  cc @n3wscott

File: kmp/diff.go [hunk 1]
```diff
@@ -36,6 +36,8 @@ func init() {
 
 // SafeDiff wraps cmp.Diff but recovers from panics and uses custom Comparers for:
 // * k8s.io/apimachinery/pkg/api/resource.Quantity
+// SafeDiff should be used instead of cmp.Diff in controllers to prevent the controller from
```
Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

Comment (vagababov) [inline kmp/diff.go]:
```suggestion
// SafeDiff should be used instead of cmp.Diff in non-test code to prevent the controller from
```

**Comment by dgerd** [comment]
Yes. There will need to be some changes to eventing and serving due to the new default diff output format. This seems to be limited to our usage of diff in the Immutability test in the webhook.

I have changes made in serving that pass units tests, and looking at what needs to change in eventing.

File: kmp/diff_test.go [hunk 2]
```diff
@@ -51,23 +50,54 @@ func TestRecovery(t *testing.T) {
...
-		t.Errorf("SafeEqual (-want, +got): %v", diff)
 	}
 }
 
-func recoveryErrorMessageFor(funcName string) string {
-	return fmt.Sprintf(
-		`recovered in kmp.%v: cannot handle unexported field: {kmp.foo}.bar
-consider using AllowUnexported or cmpopts.IgnoreUnexported`, funcName)
+func TestFieldDiff(t *testing.T) {
+	type foo struct {
```
[the remark this answers is not in the document]
Comment (dgerd) [reply kmp/diff_test.go]: Added

**Comment by dgerd** [comment]
Serving and eventing have changes that are passing tests and have been reviewed. Going to go ahead and merge this in.

/hold cancel
/lgtm

**Pull request merged**